src/test/scala/introcourse/level07/ServiceTest.scala

```diff
@@ -0,0 +1,69 @@
+package introcourse.level07
+
+import org.scalactic.TypeCheckedTripleEquals
+import org.scalatest.funspec.AsyncFunSpec
+import org.http4s.HttpApp
+import org.http4s.Request
+import org.http4s.Uri
+import cats.effect.IO
+
+class ServiceTest extends AsyncFunSpec with TypeCheckedTripleEquals {
+
+  describe("userNames") {
+    it("should return user names, separated by newlines") {
+      val usernames = Service.userNames(List(User("Vincent", 30), User("Jules", 35)))
+      assert(usernames === "Vincent\nJules")
+    }
+
+    it("should handle empty lists") {
+      val usernames = Service.userNames(List())
+      assert(usernames === "")
+    }
+  }
+
+  describe("GET /ping") {
+    it("should respond with pong") {
+      val service = HttpApp(Service.run)
+      val request = Request[IO](uri = Uri(path = "/ping"))
+
+      service(request)
+        .flatMap(response => response.as[String])
+        .map(body => assert(body === "pong"))
+        .unsafeToFuture()
+    }
+  }
+
+  describe("GET /users") {
+    it("should respond with a list of user names") {
+      val service = HttpApp(Service.run)
+      val request = Request[IO](uri = Uri(path = "/users"))
+
+      service(request)
+        .flatMap(response => response.as[String])
+        .map(body => assert(body === "Vincent\nJules\nMia\nWinston"))
+        .unsafeToFuture()
+    }
+  }
+
+  describe("GET /unknown") {
+    it("should respond with 404") {
+      val service = HttpApp(Service.run)
+      val request = Request[IO](uri = Uri(path = "/unknown"))
+
+      service(request)
+        .map(response => assert(response.status.code === 404))
+        .unsafeToFuture()
+    }
+  }
+
+  describe("usersWithLog") {
+    it("should respond with 200") {
+      val request = Request[IO](uri = Uri(path = "/unknown"))
```

Comment:
Is the `/unknown` here deliberate?

Reply:
Yes, it just needs some path that isn't covered by the pattern match in the service.

Reply:
(sorry, I just realized that this was in the wrong spec, fixed)